### PR TITLE
Refactoring required for creating extra hdd effector for Ec2

### PR DESCRIPTION
**Pull request by iyovcheva**

### CODE REVIEW

File: blockstore/pom.xml [hunk 1]
```diff
@@ -26,21 +26,6 @@
             <version>${jclouds.version}</version>
         </dependency>
         <dependency>
-            <groupId>org.apache.jclouds.provider</groupId>
```
Comment (bostko) [inline blockstore/pom.xml]: Why you excluded rackspace dependencies?

Comment (bostko) [inline blockstore/pom.xml]: There is a jclouds-rackspace specific class in the project.

File: blockstore/src/main/java/brooklyn/location/blockstore/BlockDeviceOptions.java [hunk 2]
```diff
@@ -27,7 +27,11 @@ public static BlockDeviceOptions fromMap(Map<String, ?> map) {
             }
         }
         if (map.containsKey("sizeInGb")) {
-            result.sizeInGb = (Integer) checkNotNull(map.get("sizeInGb"), "sizeInGb");
+            if (map.get("sizeInGb") instanceof Double && (Double) map.get("sizeInGb") % 1 != 0) {
```
Comment (neykov) [inline blockstore/src/main/java/brooklyn/location/blockstore/BlockDeviceOptions.java]: Comparing double to 0 with `==` could lead to surprising results. A safer alternative would be: `Math.abs(sizeInGb - (Integer)sizeInGb) < eps` where `eps` is small enough - in our case anything <=  `0.01` should work.

Perhaps include the `Float` case as well? (by casting it to double and doing the same check).

File: blockstore/src/main/java/brooklyn/location/blockstore/BlockDeviceOptions.java [hunk 3]
```diff
@@ -12,7 +12,7 @@
     private String name;
     private String zone;
     private Map<String, String> tags = Maps.newHashMap();
-    private int sizeInGb;
+    private Number sizeInGb;
```
Comment (neykov) [inline blockstore/src/main/java/brooklyn/location/blockstore/BlockDeviceOptions.java]: Why change it to a Number if you actually expect it to be an Integer (according to the error message)? Could instead cast whatever you have to int.

File: blockstore/src/main/java/brooklyn/location/blockstore/ec2/Ec2VolumeCustomizers.java [hunk 4]
```diff
@@ -69,104 +70,172 @@ public static JcloudsLocationCustomizer withNewVolumes(List<Integer> capacities)
...
     }
-    
-    public static JcloudsLocationCustomizer withNewVolumes(final Map<BlockDeviceOptions, FilesystemOptions> volumes) {
-        return new BasicJcloudsLocationCustomizer() {
-            @Override
-            public void customize(JcloudsLocation location, ComputeService computeService, TemplateBuilder templateBuilder) {
-                BlockDeviceOptions blockOptions = Iterables.getFirst(volumes.keySet(), null);
-                if (blockOptions != null && blockOptions.getZone() != null) {
-                    templateBuilder.locationId(blockOptions.getZone());
-                }
```
Comment (neykov) [inline blockstore/src/main/java/brooklyn/location/blockstore/ec2/Ec2VolumeCustomizers.java]: The customizer could get serialized, so keep the existing anonymous class - as deprecated and unused, while returning the named class instead (remove return, turning it into a regular expression, then after it do `return new NewVolumeCustomizer(volumes);`.

File: blockstore/src/main/java/brooklyn/location/blockstore/ec2/Ec2VolumeCustomizers.java [hunk 5]
```diff
@@ -69,104 +70,172 @@ public static JcloudsLocationCustomizer withNewVolumes(List<Integer> capacities)
...
      * <li>configures the EC2 availability zone</li>
      * <li>obtains a new EBS volume from the specified snapshot in the given availability zone</li>
      * <li>attaches the new volume to the newly-provisioned EC2 instance</li>
      * <li>mounts the filesystem under the requested path</li>
      * </ul>
      */
-    public static JcloudsLocationCustomizer withExistingSnapshot(final AttachedBlockDevice attachedDevice,
-            final BlockDeviceOptions blockOptions, final FilesystemOptions filesystemOptions) {
 
-        return new BasicJcloudsLocationCustomizer() {
```
Comment (neykov) [inline blockstore/src/main/java/brooklyn/location/blockstore/ec2/Ec2VolumeCustomizers.java]: Keep all those anonymous classes in.

File: blockstore/src/main/java/brooklyn/location/blockstore/ec2/Ec2VolumeCustomizers.java [hunk 6]
```diff
@@ -69,104 +70,172 @@ public static JcloudsLocationCustomizer withNewVolumes(List<Integer> capacities)
...
      * <li>mounts the filesystem under the requested path</li>
      * </ul>
      */
-    public static JcloudsLocationCustomizer withExistingVolume(final BlockDevice device,
-            final BlockDeviceOptions blockOptions, final FilesystemOptions filesystemOptions
-            ,
-            final String volumeDeviceName, final String osDeviceName, final String mountPoint,
-            final String region, final String availabilityZone, final String volumeId) {
-
-        return new BasicJcloudsLocationCustomizer() {
```
Comment (neykov) [inline blockstore/src/main/java/brooklyn/location/blockstore/ec2/Ec2VolumeCustomizers.java]: Same here.

File: blockstore/pom.xml (hunk 1 above)
Comment (neykov) [inline blockstore/pom.xml]: There are not needed for building the project, correct?

**Comment by neykov** [comment]
Some minor comments to address.

File: blockstore/src/main/java/brooklyn/location/blockstore/BlockDeviceOptions.java (hunk 3 above)
Comment (iyovcheva) [inline blockstore/src/main/java/brooklyn/location/blockstore/BlockDeviceOptions.java]: When the object is created form a map in some cases sizeInGb can be wrongly coerced to non Integer value and even if the value is of type 1.0 for example, instantiating will fail.
It is also changed for better error reporting when sizeInGb is set to double value like 2.4 for example.

Comment (neykov) [inline blockstore/src/main/java/brooklyn/location/blockstore/BlockDeviceOptions.java]: Agree, but you can convert it to an `int` on line [34](https://github.com/cloudsoft/brooklyn-blockstore/pull/3/files#diff-0fd1dffb823987019aee58245a99910cR34) (by casting).

File: blockstore/src/main/java/brooklyn/location/blockstore/BlockDeviceOptions.java [hunk 7]
```diff
@@ -74,7 +79,7 @@ public BlockDeviceOptions tags(Map<String, String> tags) {
         return this;
     }
 
-    public BlockDeviceOptions sizeInGb(int sizeInGb) {
+    public BlockDeviceOptions sizeInGb(Integer sizeInGb) {
```
Comment (neykov) [inline blockstore/src/main/java/brooklyn/location/blockstore/BlockDeviceOptions.java]: Why change to `Integer` if the `sizeInGb` field is still an `int`?

File: blockstore/src/main/java/brooklyn/location/blockstore/ec2/Ec2ExistingSnapshotCustomizer.java [hunk 8]
```diff
@@ -81,7 +81,7 @@ public void customize(JcloudsLocation location, ComputeService computeService, T
         ((EC2TemplateOptions) templateOptions).mapEBSSnapshotToDeviceName(
                 ebsVolumeManager.getVolumeDeviceName(blockOptions.getDeviceSuffix()),
                 attachedDevice.getId(),
-                blockOptions.getSizeInGb(),
+                (Integer) blockOptions.getSizeInGb(),
```
Comment (neykov) [inline blockstore/src/main/java/brooklyn/location/blockstore/ec2/Ec2ExistingSnapshotCustomizer.java]: No need for a cast?

File: blockstore/src/main/java/brooklyn/location/blockstore/ec2/Ec2VolumeCustomizers.java (hunk 5 above)
Comment (neykov) [inline blockstore/src/main/java/brooklyn/location/blockstore/ec2/Ec2VolumeCustomizers.java]: As this is new functionality, not used in the wild, then can refactor this without keeping backwards compatibility.

File: blockstore/src/main/java/brooklyn/location/blockstore/ec2/Ec2VolumeCustomizers.java [hunk 9]
```diff
@@ -69,104 +68,178 @@ public static JcloudsLocationCustomizer withNewVolumes(List<Integer> capacities)
...
+        public void customize(JcloudsLocation location, ComputeService computeService, TemplateBuilder templateBuilder) {
+            templateBuilder.locationId(blockOptions.getZone());
+        }
+
+        @Override
+        public void customize(JcloudsLocation location, ComputeService computeService, TemplateOptions templateOptions) {
+            ((EC2TemplateOptions) templateOptions).mapEBSSnapshotToDeviceName(
+                    ebsVolumeManager.getVolumeDeviceName(blockOptions.getDeviceSuffix()),
+                    attachedDevice.getId(),
+                    (Integer) blockOptions.getSizeInGb(),
```
Comment (neykov) [inline blockstore/src/main/java/brooklyn/location/blockstore/ec2/Ec2VolumeCustomizers.java]: No need for a cast?

File: blockstore/src/main/java/brooklyn/location/blockstore/BlockDeviceOptions.java [hunk 10]
```diff
@@ -27,6 +27,11 @@ public static BlockDeviceOptions fromMap(Map<String, ?> map) {
             }
         }
         if (map.containsKey("sizeInGb")) {
+            if (map.get("sizeInGb") instanceof Double && Math.abs((Double) map.get("sizeInGb") - ((Double) map.get("sizeInGb")).intValue()) >= 0.01
+                    || map.get("sizeInGb") instanceof Float && Math.abs((Float) map.get("sizeInGb") - ((Float) map.get("sizeInGb")).intValue()) >= 0.01) {
+                throw new UnsupportedOperationException("Trying to set block device with not allowed sizeInGb value "
+                        + map.get("sizeInGb") + "; sizeInGb must have integer value.");
+            }
             result.sizeInGb = (Integer) checkNotNull(map.get("sizeInGb"), "sizeInGb");
```
Comment (neykov) [inline blockstore/src/main/java/brooklyn/location/blockstore/BlockDeviceOptions.java]: If this is a `Double` you'll get a `ClassCastException`. You can convert to an `int` using casting to a `Number` and calling `intValue`.

**Pull request merged**